action/protocol/staking/contractstaking/indexer.go

```diff
@@ -82,3 +105,110 @@ func (s *Indexer) BucketTypes() ([]*BucketType, error) {
 	}
 	return bts, nil
 }
+
+// PutBlock puts a block into indexer
+func (s *Indexer) PutBlock(ctx context.Context, blk *block.Block) error {
+	// new dirty cache for this block
+	// dirty is the cache to update during event processing, which will be merged to clean cache after all events are processed.
+	// if errors occur during event processing, dirty cache will be discarded.
+	dirty := newContractStakingDirty(s.cache)
+	dirty.PutHeight(blk.Height())
+	handler := newContractStakingEventHandler(dirty)
+
+	// handle events of block
+	for _, receipt := range blk.Receipts {
+		if receipt.Status != uint64(iotextypes.ReceiptStatus_Success) {
+			continue
+		}
+		for _, log := range receipt.Logs() {
+			if log.Address != s.contractAddress {
+				continue
+			}
+			if err := handler.HandleEvent(ctx, blk, log); err != nil {
+				return err
+			}
+		}
+	}
+
+	// commit dirty cache
+	return s.commit(dirty)
+}
+
+// DeleteTipBlock deletes the tip block from indexer
+func (s *Indexer) DeleteTipBlock(context.Context, *block.Block) error {
+	return errors.New("not implemented")
+}
+
+func (s *Indexer) commit(dirty *contractStakingDirty) error {
+	batch, delta := dirty.Finalize()
+	if err := s.cache.Merge(delta); err != nil {
+		s.reloadCache()
+		return err
+	}
+	if err := s.kvstore.WriteBatch(batch); err != nil {
+		s.reloadCache()
```

Comment:
if write fails, just discard dirty cache? can we avoid reading all items from the kv? that should only be needed in Start() once